packages/devtools_server/lib/src/server.dart

```diff
@@ -92,14 +97,82 @@ void serveDevTools({
 
   final server = await shelf.serve(handler, '127.0.0.1', port);
 
+  final devtoolsUrl = 'http://${server.address.host}:${server.port}';
+
   printOutput(
-    'Serving DevTools at http://${server.address.host}:${server.port}',
+    'Serving DevTools at $devtoolsUrl',
     {
       'method': 'server.started',
       'params': {'host': server.address.host, 'port': server.port, 'pid': pid}
     },
     machineMode: machineMode,
   );
+
+  _listenForUris(devtoolsUrl);
+}
+
+void _listenForUris(String devtoolsUrl) {
+  final Stream<Map<String, dynamic>> _stdinCommandStream = stdin
+      .transform<String>(utf8.decoder)
+      .transform<String>(const LineSplitter())
+      .where((String line) => line.startsWith('[{') && line.endsWith('}]'))
+      .map<Map<String, dynamic>>((String line) {
+    line = line.substring(1, line.length - 1);
+    return json.decode(line) as Map<String, dynamic>;
+  });
+
+  // Example json input: [{"url":"ws://localhost:8888/ws"}]
+  _stdinCommandStream.listen((Map<String, dynamic> json) async {
+    final line = json['url'];
```

Comment:
change these two lines to
final uri = Uri.parse(json['url'])
Consider also adding error handling printing a message.
if (!json.contains('url']) {
  printErrorMessage(...);
}

Reply:
We should make sure to print any error messages similar to `printOutput` so they're JSON for `--machine`.